### ISSUE TITLE
[Clang][Format] Add qualifiers (actual) alignments

### ISSUE THREAD

**Issue by KBoumghar** [llvm/llvm-project#90746]
There's currently a `QualifierAlignment` option in ClangFormat, although I'd argue it is misnamed ; it doesn't align anything, it simply order qualifiers. 

Currently, I believe that if you write :

```cpp
int a = 0;
const int b = 0;
```
there is no way to obtain : 

```cpp
      int a = 0;
const int b = 0;
```

This could be added to the options of `AlignConsecutiveAssignements` and `AlignConsecutiveDeclarations`, although it'd also be useful for parameters as well, i.e. : 

```cpp
void foo(const type1 looooooooooooooooooooongarg1,
         type2 arg2){}
```

becomes : 

```cpp
void foo(const type1 looooooooooooooooooooongarg1,
               type2 arg2){}
```

**Comment by mydeveloperday** [comment]
You QualifierAlignment Left/Right/Custom, the QualifierOrder lets you control  the order... "naming is hard" but its only confusing for 5 minutes.

**Comment by mydeveloperday** [comment]
I agree this is an addition to `AlignConsecutiveAssignements`  and `AlignConsecutiveDeclarations`  maybe to suport Left/Right not just true/false

**Comment by KBoumghar** [comment]
@mydeveloperday I was thinking of maybe adding a `bool AlignQualifiers`. If this works for you (and you still believe it is a good first-time issue), I'd like to try my hand at it if I can get it assigned to me.

Any opinion on the parameter part (or does this work with declaration? I'm not sure...), this is the one my team is bugging me the most about.

**Comment by AlyElashram** [comment]
This has been stale for a while , if it's a good first issue i can take a look at it. @mydeveloperday .

**Comment by owenca** [comment]
It doesn't look like a "good first issue" to me.

**Comment by AlyElashram** [comment]
I'm still interested to take a look as I would like to actually use this type of formatting.
@owenca

**Comment by AlyElashram** [comment]
Took me a while to understand the current behavior , but I have a couple of questions.

Case 1 :
`int a = 1;
 const double b = 2;
`
should this be formatted in this way : 
`
           int a         = 1;
const double b = 2;
` 
in this way the Assignment operator still aligns and the types are aligned together. If this is the approach then we should only consider the const qualifier width and just align with it.

Another case I'd like to consider is the case of pointers. For example:
`
int *a = nullptr;
const double* b = nullptr;
`
again would it still be the same ?

`
           int *a = nullptr;
const double* b = nullptr;
`

@KBoumghar @mydeveloperday

**Comment by AlyElashram** [comment]
Took me a while to understand the current behavior , but I have a couple of questions.

Case 1 :
```
 int a = 1;
 const double b = 2;
```
should this be formatted in this way : 
```
           int a         = 1;
const double b = 2;
```
in this way the Assignment operator still aligns and the types are aligned together. If this is the approach then we should only consider the const qualifier width and just align with it.

Another case I'd like to consider is the case of pointers. For example:
```
int *a = nullptr;
const double* b = nullptr;

```
again would it still be the same ?

```
           int *a         = nullptr;
const double* b = nullptr;

```

@KBoumghar @mydeveloperday

**Comment by AlyElashram** [comment]
Took me a while to understand the current behavior , but I have a couple of questions.


Are these 2 scenarios correct for `AlignConsecutiveAssignments`
Case 1 :
```
int a = 1;
const double b = 2;
```
should this be formatted  where the i from int aligns with the d from double ? and then align the equal signs together.

![image](https://github.com/user-attachments/assets/94ac201d-4360-43d0-a274-ea43be42d1dc)

in this way the Assignment operator still aligns and the types are aligned together. If this is the approach then we should only consider the const qualifier width and just align with it.

Another case I'd like to consider is the case of pointers. For example:
```
int *a = nullptr;
const double* b = nullptr;
```

Again would we align the i of int with the d from double ?
![image](https://github.com/user-attachments/assets/1541c6c5-8a86-4370-8159-f6a6a3e4991e)

And then for `AlignConsecutiveDeclarations` we would align the variable names to be aligned together as well. Are my assumptions correct for both ?

@KBoumghar @mydeveloperday

**Comment by AlyElashram** [comment]
Just posting my findings here before unassigning this as it's a bit complicated to tackle from someone who's new to this Repo.

In `AlignConsecutiveAssignments` the function that checks for the actual _anchor_ to align is a lambda function passed to `AlignTokens`, in AlignConsecutiveAssignments this function checks for a match with the "=" sign , this behavior differs in `AlignConsecutiveDeclarations` as it searches for declarations not the equal sign. Both of these functions live inside the `WhitespaceManager`.

The suggested change would be to either call `AlignTokens` twice inside each function , once to align the qualifiers/data types and the other to align the declarations/assignments. (This is a suggestion not a proving working concept).

Other areas that require change would be `Format.h` and `Format.cpp` where you would either pass an extra variable to align qualifiers right or change the existing behaviour for `AlignConsecutiveAsssignments` and `AlignConsecutiveDeclarations` to support left and right. The suggested change would live inside `MappingTraits<FormatStyle::AlignConsecutiveStyle>`.
